### PR TITLE
Fix bug in pushing hub images

### LINKED ISSUE #786: can't push images to docker hub if image name is like "user/reponame"
```java
docker.push("mattnworb/foobar");
```

results in the DefaultDockerClient calling `RegistryAuthSupplier.imageFor("mattnworb/foobar")`. If the DefaultDockerClient is using the ConfigFileRegistryAuthSupplier, this implementation doesn't know to expand this image name and won't retrieve the config for `https://index.docker.io/v1` from ~/.docker/config.json.

### PR DESCRIPTION
**Pull request by davidxia**
when image is named "user/repo." Fixes #786

### CODE REVIEW

File: src/main/java/com/spotify/docker/client/ImageRef.java [hunk 1]
```diff
@@ -20,12 +20,17 @@
...
 import com.google.common.base.MoreObjects;
+import java.net.MalformedURLException;
+import java.net.URL;
 
 public class ImageRef {
 
   private static final String DEFAULT_REGISTRY = "docker.io";
+  private static final URL DEFAULT_REGISTRY_URL = parseRegistryUrl("https://index.docker.io/v1/");
 
+  private final URL registryUrl;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/ImageRef.java]: Decided to make this a URL.

File: src/main/java/com/spotify/docker/client/ImageRef.java [hunk 2]
```diff
@@ -75,6 +82,11 @@ public String getRegistryName() {
     return registry;
   }
 
+  /** Registry URL. */
+  public URL getRegistryUrl() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/ImageRef.java]: New method to not break backwards compat on `getRegistryName()`.

File: src/main/java/com/spotify/docker/client/ImageRef.java [hunk 3]
```diff
@@ -83,4 +93,23 @@ public String toString() {
         .add("tag", tag)
         .toString();
   }
+
+  /** Return registry server address given first part of image. */
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/ImageRef.java]: Is there a better way?

File: src/main/java/com/spotify/docker/client/ImageRef.java [hunk 4]
```diff
@@ -75,6 +80,11 @@ public String getRegistryName() {
     return registry;
   }
 
+  /** Registry URL. */
+  public String getRegistryUrl() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/ImageRef.java]: New method to not break backwards compat on `getRegistryName()`.

File: src/main/java/com/spotify/docker/client/ImageRef.java [hunk 5]
```diff
@@ -83,4 +93,23 @@ public String toString() {
...
+      return DEFAULT_REGISTRY_URL;
+    } else if (url.matches("(^|(\\w+)\\.)gcr\\.io$") || url.equals("gcr.kubernetes.io")) {
+      // GCR uses https
+      return "https://" + url;
+    } else if (url.equals("quay.io")) {
+      // Quay doesn't use any protocol in credentials file
+      return "quay.io";
+    } else if (!url.contains("http://") && !url.contains("https://")) {
+      // Assume http
+      return "http://" + url;
```
Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/ImageRef.java]: why assume http? I think https is probably a smarter assumption to make.

Comment (mattnworb) [inline src/main/java/com/spotify/docker/client/ImageRef.java]: Docker will interpret a "registry URL" like `example.com/foo/bar` as a registry that uses `https` and will only pull over `http` if specially configured to do so: https://docs.docker.com/engine/reference/commandline/pull/#pull-from-a-different-registry

**Comment by mattnworb** [comment]
I think the root bug fix - knowing when to use `index.docker.io` for docker hub images - is correct, but I think the default of http is dangerous

**Comment by davidxia** [comment]
Updated to use https.

**Comment by mattnworb** [comment]
:+1:

**Comment by davidxia** [comment]
hopefully we won't need to add more custom registry names to `ImageRef`.

**Pull request merged**